txcache/selection.go

```diff
@@ -0,0 +1,69 @@
+package txcache
+
+import "container/heap"
+
+func (cache *TxCache) doSelectTransactions(gasRequested uint64, maxNum int) (BunchOfTransactions, uint64) {
+	senders := cache.getSenders()
+	bunches := make([]BunchOfTransactions, 0, len(senders))
+
+	for _, sender := range senders {
+		bunches = append(bunches, sender.getTxsWithoutGaps())
+	}
+
+	return selectTransactionsFromBunches(bunches, gasRequested, maxNum)
+}
+
+// Selection tolerates concurrent transaction additions / removals.
+func selectTransactionsFromBunches(bunches []BunchOfTransactions, gasRequested uint64, maxNum int) (BunchOfTransactions, uint64) {
+	selectedTransactions := make(BunchOfTransactions, 0, initialCapacityOfSelectionSlice)
+
+	// Items popped from the heap are added to "selectedTransactions".
+	transactionsHeap := make(TransactionsMaxHeap, 0, len(bunches))
+	heap.Init(&transactionsHeap)
+
+	// Initialize the heap with the first transaction of each bunch
+	for i, bunch := range bunches {
+		if len(bunch) == 0 {
+			// Some senders may have no eligible transactions (initial gaps).
+			continue
+		}
+
+		// Items will be reused (see below). Each sender gets one (and only one) item in the heap.
+		heap.Push(&transactionsHeap, &TransactionsHeapItem{
+			senderIndex:      i,
+			transactionIndex: 0,
+			transaction:      bunch[0],
```

Comment:
so you only select one here?

Reply:
First, we gather all transactions of all senders in `bunches`. Each bunch (sourced from a single sender) is nonces-gap-free (though duplicates weren't properly handled, fixed that during this fix after review :see_no_evil:) and the containing transactions are nicely sorted by nonce.

Secondly (still in the "setup" phase), we pick the "head" of every bunch - in other words, we put the first transaction of each sender on a max heap.

Afterwards, in a loop, until we reach our target (gas or count), we pop the best transaction from the heap and append it to the slice of `selectedTransactions`; then, we immediately pick the next transaction of the respective bunch (whose transaction was _selected_) and place it on the heap to compete with the existing ones.

Reply:
ok. this explanation needs to be put somewhere.

Reply:
Added here:
 - https://github.com/multiversx/mx-chain-storage-go/blob/selection-by-ppu/txcache/README.md#paragraph-4